### ISSUE TITLE
Array method definition revamp: Use case collection

### ISSUE THREAD

**Issue by RyanCavanaugh** [microsoft/TypeScript#36554]
We've gotten numerous issue reports and PRs to change the methods of `Array`, particularly `reduce`, `map`, and `filter`. The built-in test suite doesn't cover these very well, and these methods interact with each other and the surrounding contextual type in fairly subtle ways.

@jablko has done a great job at #33645 collecting a variety of issues into a single PR; we need to augment this PR (or something like this) with a proper test suite so we can be sure about what's being changed.

I'd like to create a clearinghouse issue here to collect **self-contained** (I CANNOT POSSIBLY STRESS THIS ENOUGH, **SELF-CONTAINED**, DO NOT IMPORT FROM RXJS OR WHAT HAVE YOU) code samples that make use of the array methods.

Please include with your snippet:
 * Compiler settings
 * Compiler version you were testing with
 * The **expected** behavior (examples that should and should not produce errors are both useful)
 * No imports or exports; snippets need to be self-contained so that we can put them into our test suite without extra complications

Once we've established a critical mass of code snippets, we can start combining the existing PRs into an all-up revamp and assess its impact to real-world code suites to figure out which changes don't result in unacceptable breaking changes.

ｓｅｌｆ－ｃｏｎｔａｉｎｅｄ

**Comment by RyanCavanaugh** [comment]
From #19535: `concat` should reflect the flattening of its input argument
```ts
const foo: [number, string][] = [[1, 'one']];
const a = foo.concat([2, 'two']);
// SHOULD ERROR,
// the actual content of 'a' is [[1, 'one'], 1, 'two']
// so a[1] should be number | string | [number, string]
a[1][0];
```

**Comment by RyanCavanaugh** [comment]
From #24579: `flat` should, well, flatten. Tested on 3-8 beta with target: ESNext, strict on
```ts
declare let arr: [[number, boolean], string];
let x0 = arr.flat(0); // Should be [[number, boolean], string] or (number | boolean | string)[]
let x1 = arr.flat(1); // Should be [number, boolean, string] or (number | boolean | string)[] 
let x2 = arr.flat(2); // Should be [number, boolean, string] or (number | boolean | string)[]
```

**Comment by RyanCavanaugh** [comment]
From #26976: `concat` should at least allow an empty array as a target. Arguably it should allow heterogenous operations? Tested on 3.8-beta with target: ESNext, strict on
```ts 
// Should be OK (currently an error) and produce string[]
let a1 = [].concat(['a']);
// Should be OK (is) and continue to produce string[]
let a2 = ['a'].concat(['b']);

// Should either error (current behavior) or maybe produce (string | number)[]
let a3 = [1].concat(['a']);
```

**Comment by RyanCavanaugh** [comment]
From #29604: `flat` shouldn't produce `any[]` (?!), `Array.prototype.concat` should be better if possible. Tested on 3.8-beta with target: ESNext, strict on
```ts
// Should be an error
const a: boolean[] = [[17], ["foo"]].flat();
// Should be an error (stretch goal)
const b: boolean[] = Array.prototype.concat([17], [19], [21]);  
```

**Comment by RyanCavanaugh** [comment]
From a real-world code suite broken by #33645:
```ts
// Should not error
function add<A>(arr: Array<A>, el: A): Array<A> {
    return arr.concat(el)
}
```

**Comment by RyanCavanaugh** [comment]
From a real-world code suite broken by #33645:
```ts
// Did not error; probably shouldn't in the future either
class A {
    flattenTree(option: any, changeOnSelect: any, ancestor = []) {
        let flattenOptions: any = [];
        const path = ancestor.concat(option);
        flattenOptions = flattenOptions.concat(this.flattenTree(option, changeOnSelect, path));
   }
}
```

**Comment by falsandtru** [comment]
```ts
([] as any[]).reduce(() => 0, 0); // Expected: number, Actual: any
([] as unknown[]).reduce(() => 0, 0); // Expected: number, Actual: unknown
([] as never[]).reduce(() => 0, 0); // Expected: number, Actual: number
```

**Comment by falsandtru** [comment]
@RyanCavanaugh This is caused by a mistake of the order of overloads. I made #36570 and looks like it makes no regression.

**Comment by jablko** [comment]
> From a real-world code suite broken by #33645:
> 
> ```ts
> // Should not error
> function add<A>(arr: Array<A>, el: A): Array<A> {
>     return arr.concat(el)
> }
> ```

@RyanCavanaugh Shouldn't it? If `A` is `string[]`, `arr` is `[["a"]]` and `el` is `["b"]` then `add(arr, el)` will return `[["a"], "b"]`, which doesn't match the return type `Array<A>`?

Should that implementation be updated -> `return arr.concat([el])`?

**Comment by RyanCavanaugh** [comment]
@jablko good point. The actual code was in `fp-ts` and indeed they've removed it; the implementation now uses a loop instead https://github.com/gcanti/fp-ts/blob/master/src/Array.ts#L373

**Comment by jablko** [comment]
> From a real-world code suite broken by #33645:
> 
> ```ts
> // Did not error; probably shouldn't in the future either
> class A {
>     flattenTree(option: any, changeOnSelect: any, ancestor = []) {
>         let flattenOptions: any = [];
>         const path = ancestor.concat(option);
>         flattenOptions = flattenOptions.concat(this.flattenTree(option, changeOnSelect, path));
>    }
> }
> ```
> 
> This one should be simpler to demonstrate exactly what aspect of things got broken

@RyanCavanaugh I think this should be an error? The reason it currently doesn't error is because `option: any` is assignable to `ConcatArray<never>` (at `const path = ancestor.concat(option)`). However the only way the return type `never[]` is accurate is if `option` is `[]`.

I think the `flattenTree()` signature needs to be `ancestor: any[] = []`, to be callable with the `const path = ancestor.concat(option)`, where `option` can be something other than `[]`? That, or the signature must be `option: never[]`.

**Comment by bpasero** [comment]
**TypeScript Version:**  3.8.x

**Code**

```ts
interface Bar {
    property: boolean;
}

let foo: ReadonlyArray<string> | Bar;
if (Date.now() === 0) {
    foo = { property: true};
} else {
    foo = ['Hello'];
}

if (Array.isArray(foo)) {
    console.log(foo[0]);
} else {
    console.log(foo.property); // <-- error
}
```

**Expected behavior:**
The else branch correctly understands that `foo` is of type `Bar`

**Actual behavior:**
TypeScript still thinks foo is `ReadonlyArray<string> | Bar`

**Playground Link:** [link](https://www.typescriptlang.org/play/#code/JYOwLgpgTgZghgYwgAgEJysg3gKGf5ABygHtDowBPALmQCMSSAbCOEAbhwF8ccWxkMRrQBKrACYkQTSgEEoUOJQA8AZzBRQAcwB8yAD5oMnYDGQAKACJxIAOhAkA7uYCUyALyfkABje4Cgowe2ESk5FBUtBoArhBcnFzIEEyqKP4EQiTBANoA5AASyUwkuQC6CbymFvKKlLbAqjVK5pkufngECFKqzBC2xVotjNnepS4JSSlpHfhdID0s-SSDmbbEZBSU49xAA)

**Comment by HolgerJeromin** [comment]
@bpasero This has nothing to do with Array stuff:

	let foo: boolean | string;
	if (Date.now() === 0) {
		foo = true;
	} else {
		foo = "hello";
	}
	foo; // still of type: boolean | string

Not possible without defining a specialized type "positive number" (no zero and even no negative value, no decimal points)

**Comment by bpasero** [comment]
@HolgerJeromin I just did what @DanielRosenwasser suggested in https://github.com/microsoft/TypeScript/issues/37129#issuecomment-593661062 and posted my usecase here.

**Comment by malthejorgensen** [comment]
A lot of people use `Array.reduce()` to build mappings from lists:

```
type MyMapping = { [id: string]: 'A' | 'B' | 'C' }

const aMap: MyMapping = [1, 2, 3].reduce((acc, n) => {
  return { ...acc, [`id_${n}`]: 456 }
}, {})
```

I would expect this to not typecheck since `456` isn't of the type `'A' | 'B' | 'C'`.

This currently typechecks with no errors in Typescript v3.8.3.

**Comment by xiaoxiangmoe** [comment]
```ts
const a = [[1, 2, 3]].flat(); // number[]
const b = ([[1, 2, 3]] as const).flat(); // unknown[]
```

**Comment by xiaoxiangmoe** [comment]
Also , we need Infinity literal type. Here is an example we use:

```ts

type Element<T> = {
  isArray: ArrayElem<T extends ReadonlyArray<infer E> ? E : never>;
  notArray: T;
}[T extends ReadonlyArray<any> ? 'isArray' : 'notArray'];

type ArrayElem<T> = T extends ReadonlyArray<any> ? Element<T> : T;

declare function flatInfinity<T>(x: T): Array<ArrayElem<T>>;

const c = flatInfinity(['foo', 1, 2, null, [5, [2], [3, {}, 4]]] as const); // ("foo" | 1 | 2 | 5 | 3 | 4 | {} | null)[]
```

**Comment by HoldYourWaffle** [comment]
`map` erases "tuple-ness" of an array (#29841):
```ts
type Vec3D = [number, number, number];
let vec: Vec3D = [1, 2, 3];
let scaledVec: Vec3D = vec.map(x => 2 * x); // error: Type 'number[]' is missing the following properties from type '[number, number, number]': 0, 1, 2
```

The following overload using mapped tuples can be defined as workaround (as seen in #29841):
```ts
declare interface Array<T> {
    map<U>(callbackfn: (value: T, index: number, array: T[]) => U, thisArg?: any): { [K in keyof this]: U };
}
```

**Comment by proteriax** [comment]
Suggestion:

```ts
declare global {
  interface Array<T> {
    filter(filter: BooleanConstructor): Exclude<T, null | undefined | 0 | "" | false>[]
  }
}
```

**Comment by jdmoody** [comment]
In addition to @HoldYourWaffle's `map` use case, it would be great to support generic tuples of arbitrary arity:
```typescript
interface Box<T> { value: T }
type BoxTuple<T> = { [K in keyof T]: Box<T[K]> }

const forBoxes = <T extends any[], TValue> (...boxes: BoxTuple<T>) => ({
  useValues: (callback: (v: T) => TValue) => {
    // Argument of type 'any[]' is not assignable to parameter of type 'T'.
    //   'any[]' is assignable to the constraint of type 'T',
    //   but 'T' could be instantiated with a different subtype of constraint 'any[]'.
    return callback(boxes.map((box) => box.value))
    // (Asserting the map call 'as T' fixes the error)
  }
})

const values = forBoxes(
  { value: 'Dog' },
  { value: 1 },
  { value: true }
).useValues(
  ([str, num, bool]) => `My boxes contain ${str}, ${num}, and ${bool}`
)
```
([Playground Link](https://www.typescriptlang.org/play/?ssl=7&ssc=62&pln=9&pc=51#code/JYOwLgpgTgZghgYwgAgEIHsAeAeAKgPmQG9kA3OAGwFcIAuZXZAXwCgwBPABxQ01ys4UIeQgF5iyANoBpZKGQBrCO3QwGAXXq88M9YVYsE6EAGcwyGOii8IJ5OLzIImSCAAmduCHaT1AGgYANUoaQgAKADoogCMsWy0sfkFhAgBKe3CiFmRkKhMIYOp45DCESgpoxAV6MNJ6XHTRQlxCmkbCLJycgHpu5ABBKABzKgBbCHBkVWQObmQAci8fdXm5OxB0czgTE2AhkDhooRn0ZE44KDhxyCgptVmUedx5iOyu3q7F719V4E8dvYHI4oMCnMAACxQRlMYEuoHM0weC2efjePT6OWiVHMT1WRioFDcyGiKFAZi8YGAcEgRIA7sAIcg4Mg3MAYDBoBNzCYqNEkdNoWY4ZMvssXmjkFAIGAqFAQMgyhQKlUwrFMLYIqM4JwwqqsO1iVgIuQiqlUhKPmF+jtoJSQEMZpDkFrOArygttgxVjBgOq7BCULguChoFArOacqwmObDMYzGQQrZ7BYrDYTGE3iQTTR6PMACLoIarJionJZxP0ACMzFLEmzdBmUBozBYqQieQKifTbzCkiFARAYwCsXQFHUBoABgBZdiGv0K4xgODyAAkRCFJeQa8Ho03XiJa5HFCYE9bLFjplHEAiFELtS7qSAA))

**Comment by extmchristensen** [comment]
I would like concat.apply in _getEvents to work without the typecast of [] to ReadonlyArray<string>.

------
```
export function _getEvents(
  nameSpec: string | RegExp | Array<string | RegExp>
): ReadonlyArray<string> {
  if (Array.isArray(nameSpec)) {
    const allStrings = ([] as ReadonlyArray<string>).concat.apply([], nameSpec.map(a => _getEvents(a)));
    const allUniqueStrings = Array.from(new Set<string>(allStrings).values());
    return [
      ...allUniqueStrings
    ];
  } else {
   // ...
  }

  return [];
}
```

**Comment by jimador** [comment]
I created #39259 . It appears that 2 of the signatures for `reduce` are colliding, which breaks inference for the return type of `reduce` in certain situations. It appears that the conflicting signatures are: 

```ts
    reduce(callbackfn: (previousValue: T, currentValue: T, currentIndex: number, array: readonly T[]) => T, initialValue: T): T;
    reduce<U>(callbackfn: (previousValue: U, currentValue: T, currentIndex: number, array: readonly T[]) => U, initialValue: U): U;
```

It looks like the first can be dropped in favor of the second, as signature of the second covers the first.

[Here](https://www.typescriptlang.org/play/?ssl=31&ssc=2&pln=1&pc=1#code/C4TwDgpgBAyg9gWwgBQIYCdUICrmgXigG8oALCAGwrgC4oBnYdASwDsBzAbigHc50KAEzqMWHbgF8AsAChZAMwCurAMbBmcVlABG6OAGsIrAJKt5EdEZUQAPNgB8ACnqIIAQXTs6RANoB9GlE2dgBdESZgiQBKOm04OAoIVC0iWSgoUEgoAGkIEFwswkMQOHkMvFKGVw92NKgVTUYoYvooQkcAeW0AKwg1ADoW52rPKKhUVtz8vB8QqNk6y2BFdC0W-stBRWtHRzBLADcNRXoANVQKRQgAGnqVy1Zgc8uIMfx7Yjr09IbWJst6IoKMA2nd0A8nhcrm1CABychUOC8fhCWFfb5LFZafYQI5wE7PaEAMiJUABQOA6IktyYV3mMmkchkSlU6k09X4ljUpnMDx2LiQNTo8CQaEwODwMR08USyU+MnSmWgUwKBGaeUqSsqAvcnjqvyaLVBnR6fWAgzy9GGgtG40meVVs3p6UWEGWq3VIHoGwgW2sNjiCSSrCce0OxzOUJuYIhhNebQ+qQV3x+jRBwAgTUIOpqPhU9yMkJeIRhcIR1GRAkEaOTKcxHpxeIJUagJIymcptag1Iy6DpskZCmUag0Wga4LNPIsVggAEZrbqvLBXGKsKqpYHZSk6kqcg68KDipqKmUc3rkwaQUb2l1egMhmf2GMJnvppAnfq01ADi2ipafX6EC7I2EZxrc+YTo8cZvIm6Kpn86YdqCj55gWUFRiW+BYVA8KUBWfBVjWKbpPW2LhvikYvK2pIZowVI0n2ryuu6Wg-i8A5AA) is a playground link demonstrating the issue.

**Comment by julian-sf** [comment]
@jimador and I were looking at this one together.  Here is an even shorter repro path ([playground link](https://www.typescriptlang.org/play/?ssl=1&ssc=1&pln=23&pc=2#code/C4TwDgpgBAyg9gWwgBQIYCdUICrmgXigG8oALCAGwrgC4oBnYdASwDsBzAbigHc50KAEzqMWHbgF8AsAChZAMwCurAMbBmcVlABG6OAGsIrADzYAfAAoM7OvCRpMOPAEo62uHAoRUWorKhQ6BDAiuhaAPLaAFYQagB0hiD0VujsznFBgooqEBb+AVAWYEEAbhqK9ABqqBSKEAA0UCqhQazA1bUQzlD4ZlDFEGVwFR11UABk41AAhNPN6K3tNXX1+QFMdfnOstJyMkqq6pq8-PrJ1raIKBhYuJCuOh5ePsT5QSFhUJEx8YnnqelMtkIMZ3J5vKxLGtCgMhiNlg0mi0jEtOt1ev1SuUqgiJlNZvNFqMGtCNhAtjtZAplGoNFoavQ4AB1U7-GywK4OW4uNxPCGvGQBFSaRhQEq4wjfWLABIQJIpNIZCBZHJ5QUFIpY4Y4zqNQko4novqw7HEvEzObItrE1bq9boTbq7bq96hLTizo7IA)):

```ts
type SomeParamType = { hello: string; world: string; }

function broken<T>(arg: SomeParamType): boolean {
  // This is what's throwing the TypeError, seems like an inference issue here
  return Object.keys(arg).reduce( 
    (previousValue, currentValue) => previousValue && !!currentValue,
    true
  )
}

function works(arg: SomeParamType): boolean {
  return Object.keys(arg).reduce<boolean>(
    (previousValue, currentValue) => previousValue && !!currentValue,
    true
  )
}

function alsoWorks(arg: SomeParamType): boolean {
  const value = Object.keys(arg).reduce(
    (previousValue, currentValue) => previousValue && !!currentValue,
    true
  )
  return value
}
```

**Comment by jimador** [comment]
> I'm very interested to know what mechanism causes the first and third (broken and alsoWorks) examples to behave differently.

This has piqued my interest as well.

**Comment by jimador** [comment]
@julian-sf That's interesting, it looks like TS is downcasting (or just plain inferring) the return type as `true` and not `boolean`. If it's an inference problem, this should be reproducible.

**Comment by AndrewMorsillo** [comment]
Perhaps I'm missing something here, I'm not super experienced with more complex types, but this case looks impossible with TS right now.

I have a const array of strings. I want an object with those strings as keys.

```
const myKeys = ["foo", "bar"] as const
type TMyTypeWithMyKeys = {
    [key in typeof myKeys[number]]: string
}

//Type '{}' is missing the following properties from type 'TMyTypeWithMyKeys': foo, bar
const myObjWtihMyKeys: TMyTypeWithMyKeys = myKeys.reduce((obj, key) => ({ ...obj, [key]: "initialValue" }), {})
```

Desired behavior is for TS to infer the return type instead of just using the initial value as the return type.

**Comment by jimador** [comment]
@AndrewMorsillo,

The signature for `reduce` declares the return type of reduce as the type of the initial value (`{}` in this case). TS would have to execute the reduce function to infer the return type (kind of like a const exp?). If TS _could_ do something like that, the type of the result would probably be `{ foo: 'initialValue'; bar: 'initialValue'}` which still wouldn't match `{ foo: string; bar: string}`

**Comment by julian-sf** [comment]
> Perhaps I'm missing something here, I'm not super experienced with more complex types, but this case looks impossible with TS right now.
> 
> I have a const array of strings. I want an object with those strings as keys.
> 
> ```
> const myKeys = ["foo", "bar"] as const
> type TMyTypeWithMyKeys = {
>     [key in typeof myKeys[number]]: string
> }
> 
> //Type '{}' is missing the following properties from type 'TMyTypeWithMyKeys': foo, bar
> const myObjWtihMyKeys: TMyTypeWithMyKeys = myKeys.reduce((obj, key) => ({ ...obj, [key]: "initialValue" }), {})
> ```
> 
> Desired behavior is for TS to infer the return type instead of just using the initial value as the return type.

Technically you can achieve it with this:

```
const myKeys = ["foo", "bar"] as const
type TMyTypeWithMyKeys = {
    [key in typeof myKeys[number]]: string
}

const myObjWtihMyKeys = myKeys.reduce<TMyTypeWithMyKeys>((obj, key) => ({ ...obj, [key]: "initialValue" }), {} as any)
```

**Comment by nigelsim** [comment]
In Typescript 3.9.7.

```
function toStringArray(a: string[] | number[]) {
    return a.reduce((acc, v) => acc.concat(v.toString()), [] as string[]);
}
```
`v` should be inferred as `string | number` but ends up as `any`. Likewise, `acc` should be `string[]` but is `any`.

`reduce` has the error
> Each member of the union type '{ (callbackfn: (previousValue: string, currentValue: string, currentIndex: number, array: string[]) => string): string; (callbackfn: (previousValue: string, currentValue: string, currentIndex: number, array: string[]) => string, initialValue: string): string; <U>(callbackfn: (previousValue: U, currentValue: string, ...' has signatures, but none of those signatures are compatible with each other.ts(2349)

**Comment by ExE-Boss** [comment]
I believe that `.map(…)` operations on tuple types should still produce a tuple type.

**Comment by asbjornh** [comment]
Typescript version 4.0.5

This one is a little longer than other examples here (I'm sorry), but I haven't been able to recreate this with a simpler function.

The behaviour is very interesting though. Type inference for my `groupBy` utility works as I'd expect as long as the type of list elements don't have properties that are of type `Array`.

```ts
const groupBy = <K extends string, V>(func: (el: V) => K) => (
  acc: Record<K, V[]>,
  el: V
): Record<K, V[]> => {
  const key = func(el);
  return { ...acc, [key]: (acc[key] || []).concat(el) };
};

type A = { a: string; b: number };
const listA: A[] = [];
// 'el' is inferred to be of type 'A' as expected
const a = listA.reduce(groupBy(el => el.a), {});

type B = { a: string; b: string[] }; // Note that 'b' is a list
const listB: B[] = [];
// 'el' is inferred to be of type 'string | B'
const b = listB.reduce(groupBy(el => el.a),  {}); // Type error:  Property 'a' does not exist on type 'string | B'
```

The `el` argument is also inferred to be `string | T` if `T` has properties of type `Record` or indexed types (and maybe others)

I also found that manually deleting the following override from `lib.es5.d.ts` makes inference work as expected for both cases:
```ts
    reduce(callbackfn: (previousValue: T, currentValue: T, currentIndex: number, array: T[]) => T, initialValue: T): T;
```

**Comment by Retsam** [comment]
With `.split`, it'd be nice if the return type reflected that the method *always* (with two caveats) returns an array with at least one string in it.  So the return type could be `[string, ...string[]]` instead of `string[]`.  

This would make code like this function better with the `noUncheckedIndexedAccess` flag:

```ts
// currently head is `string | undefined`, but in practice, it's always `string`
const [head, ...rest] = someString.split("/");
```

There are, AFAIK, two caveats:

1. `someString.split(sep, 0);` will return `[]`, because to the `limit` being `0`.
2.  `"".split("")` returns `[]`  not `[""]`, because... "Life is pain, highness; anyone who tells you differently is selling something" or something.

Technically, those could be addressed with overloads, but at least the second one feels like a reasonable sacrifice of soundness for the sake of pragmatism.

**Comment by leavesster** [comment]
TypeScript 4.1.2

except give a ts error, such as `Type 'string' is not assignable to type 'string[]'.ts(2322)`

```typescript
let list:string[][] = [["a"]];
list = list.concat(["l"]);
// SHOULD ERROR, the actual content of list is 
// [ [ 'a' ], 'l' ];
```

**Comment by samhh** [comment]
@Retsam If we had a nice way to define a non-empty string (like we can with arrays) then you could address the soundness issue by overloading on non-empty input.

**Comment by icecream17** [comment]
Here, I have to use `slice` since `_2Darray` is readonly. But then type information is lost and there's an error
```ts
const _2Darray = [
    [1, 2],
    [3, 4]
] as const

function getX(a: number): [number, number] {
    return _2Darray[a].slice()
}
```

```ts
Type '(1 | 2)[] | (3 | 4)[]' is not assignable to type '[number, number]'.
  Type '(1 | 2)[]' is not assignable to type '[number, number]'.
    Target requires 2 element(s) but source may have fewer.
```

```ts
_2Darray[a]: readonly [1, 2] | readonly [3, 4]
// I would expect `_2Darray[a].slice` to be:
[1, 2] | [3, 4]
```

**Comment by tjjfvi** [comment]
@icecream17 `[..._2Darray[a]]` works, fwiw

**Comment by shadow-light** [comment]
Related to above, I'd expect `slice()` to clone an array and keep its type, but it doesn't keep type when min number of items is specified.

```typescript
const array_one_or_more:[number, ...number[]] = [1, 2, 3]
const clone = array_one_or_more.slice()
clone[0].toString()  // Error because clone has become `number[]`
```

**Comment by Asaf-S** [comment]
Related to the above, this old issue [#33200](https://github.com/microsoft/TypeScript/issues/33200) wasn't solved yet, instead, a workaround was offered, but it creates discomfort.

I've created the following PR to fix it (also, a better explanation about the problem was mentioned here): [PR 45767](https://github.com/microsoft/TypeScript/pull/45767).

**Comment by kwasimensah** [comment]
Adding an interaction between flat and Promises

https://www.typescriptlang.org/play?target=8#code/IYZwngdgxgBAZgV2gFwJYHsI1JKAxAG2GQFlgAHAHgGUAaGAFQD4AKAKBhgDdgCEBTEAC4YAJX7AAJpgJgAggCcFwMDSa0OMALYURLHn34jqAShgBeJjAAKC9FtQh+lBjAA+YidIizFy1cxMbCYitvaOzuJSMvJKKi5MVgDemlCYIMgw5HYOTiAW3LwCIAB0OuQs5SYA3KnpmYgEcKgEBPySAGpFggXAAO7AqJlhufwlvAQs2eF5NZoK-MgICliNza3tXYalcETILHMAvkA

```
async function asyncFlatMap<S, T>(
  values: ReadonlyArray<S>,
  map: (value: S) => Promise<T | ReadonlyArray<T>>
): Promise<ReadonlyArray<T>> {
  const promises = values.map(map);
  const fulfilledValues = await Promise.all(promises);
  return fulfilledValues.flat(); // Type system doesn't realize this should be Array<T>
}
```

**Comment by tjjfvi** [comment]
@kwasimensah That's a problem with the typing of the external function; `T` could be, for example, `number[]`, in which case the `.flat()` would return a `(number | number[])[]`. You should instead type it like `flat`:
```ts
async function asyncFlatMap<S, T>(
  values: ReadonlyArray<S>,
  map: (value: S) => Promise<T>
): Promise<(T extends ReadonlyArray<infer U> ? U : T)[]> {
  const promises = values.map(map);
  const fulfilledValues = await Promise.all(promises);
  return fulfilledValues.flat();
}
```

**Comment by lazytype** [comment]
I think we could introduce changes to some of these methods in a backwards-compatible way. Here's a definition of `concat` that I believe satisfies this:
```ts
declare interface Array<T> {
    concat<U>(this: Array<U>, ...items: ConcatArray<U>[]): Array<U>;
    concat<U>(this: Array<U>, ...items: (U | ConcatArray<U>)[]): Array<U>;
}

[1].concat('x')  // error as expected
[1].concat<number | string>('x')  // No error since type is explicitly widened
```
Playground: https://tsplay.dev/wXkDJW

I think this approach should work for most the Array methods in question.

**Comment by vjau** [comment]
I don't know if this is appropriate here, but i have just been bitten by this.

Imho, ```filter()```, ```some()```, ```every()```... should only accept a predicate returning a boolean.

```ts
//strict compiler settings

const arr = [1, 2, 3];


const onlyPositives = arr.filter(num=>{
  num > 0 //forgot return, TS doesn't mind
});

console.log("onlyPositives", onlyPositives); // []  empty array


//isPositive has type (num:number)=>void
const isPositive = (num:number)=>{
  num > 0;
}

//filter should not accept a predicate not returning a boolean
const onlyPositives2 = arr.filter(isPositive);//no error that way either
```

Instead, the predicate type is actually returning ```unknown``` in the definitions. 

Sorry if this has already been discussed to death, but i find this very unsound and i havn't been able to find a duplicate.

**Comment by tjjfvi** [comment]
> Imho, `filter()`, `some()`, `every()`... should only accept a predicate returning a boolean.

This would be a pretty big breaking change, as many people will use e.g. `.filter(x => x)` to remove falsy values (e.g. to remove `null` or empty strings).

Perhaps, instead, the type could be modified in some way to allow `unknown` but ban `void`? Perhaps an overload expecting a void return type with `@deprecated You probably meant to return something`?

**Comment by Retsam** [comment]
@vjau You can also add the [@typescript-eslint/no-unnecessary-condition](https://github.com/typescript-eslint/typescript-eslint/blob/v4.9.0/packages/eslint-plugin/docs/rules/no-unnecessary-condition.md) lint rule - it tries to find dead conditionals and checks these array predicates.

**Comment by vjau** [comment]
> This would be a pretty big breaking change, as many people will use e.g. `.filter(x => x)` to remove falsy values (e.g. to remove `null` or empty strings).

I was under the impression that it was idiomatic TS to write `.filter(x => !!x)` instead but after some research it seems i am probably confusing with another language.

**Comment by Brookke** [comment]
Hey @RyanCavanaugh, please could I get https://github.com/microsoft/TypeScript/issues/41708 added to this so that #44216 can finally get merged/fixed.

**Comment by EthanRutherford** [comment]
[playground link](https://www.typescriptlang.org/play?#code/MYGwhgzhAEAiCmwDW14A8Au8B2ATGAggE5FgCe0A3gFDTQQAWArgGYsjwAUAlFbXdAD0g+szYdowAPa540AOZS5DeEXj8Avv2HQMK6GoxMi2XWQAOcgJYwwcRCjDBpRXFezzdU6ACkwAN0hgIitzDAAaaAAjJgwFeAwYDAt4XGhIaGJSCiiKZMsIYNCMflxSAHdOaSZsDAAuaGwmAFso1V4aAEhOw2NTPRsAOghzECtgLgBaatrImYxuAG5qTq06HQTgQeoNIA)

Array methods such as splice, map, etc. when called on a subclass of Array, return a new instance of the subclass.

This might be related to Tuples losing their Tuple-ness when calling map, etc., but I'm not sure if that's the same issue. Tuples exist only in the type system, while a subclass exists at runtime, so I could see this potentially being a separate issue from the tuple issue.

**Comment by EthanRutherford** [comment]
Ah, I was not aware of that.

**Comment by EthanRutherford** [comment]
After looking further into https://github.com/tc39/proposal-rm-builtin-subclassing, it appears that support for "Type II: subclass instance creation in built-in methods" support is unlikely to be possible to remove without significant webcompat issues. One of the champions has even responded that the removal of type II is highly unlikely: https://github.com/tc39/proposal-rm-builtin-subclassing/issues/21#issuecomment-851566785.

In light of that, and the fact that the proposal is only stage 1, perhaps it would be prudent for Typescript to try to match the current behavior, rather than anticipate what seems to be a rather unlikely future behavior change?

**Comment by ackvf** [comment]
`lib.es2016.array.include.d.ts` should not produce error when testing existence of an element that comes from a wider set of values that intersects with `Array<T>`.

[playground link](https://www.typescriptlang.org/play?#code/CYUwxgNghgTiAEYD2A7AzgF3hkmCMAXPHgFCiSwLLpY6YBMRe8APvPWeNHIqptrgwBmIvVbwhnCj2r86GACxEUAVwC2AIxAwSU7lT5ZYMKAE8iACgAM4vAEoA2gF1dxswDoAlikgrQaC3l7eAB6EPgAeQBpEjdTLx8IP1xAwXo7UPCAQRgAc3UQFCwkADNsUwAHBAByZjZ6avhPNHgUJCM0NE9clCgNCAQMJHgK2Cg1EBwYeFLyqvhqmzY8avd4ABVKmoamlraOrp6+gexhjC2FpeJVi3ohBQBWO1iYE3jvX39UzCEMsPgcvkJkUZmVzvNqmI2EJGs1Wu14FBOt1ev1BsNRiYJlNQXMalcVmtNhCdnD9ojkUc0ac8ZdbDc7o9nnEEp8UvIFH9snkCiDZuCaqpNNpYXsEUjDqiTkMRmNsdpcQK6csGfcnq5Xh4St5gABJFCgAAeFgsADcMgBeAB88FN8AtDoE+C5kRiLO1Bv1RpN5vtNrtDotTow6Uyrpeb3cHr1BpAxrNlv99sd8l+YfWAAs4dRgJ4MJ5UPAAO6eCAQREQItmFpwDAqGAoBYlKAQNAgRpdHyDDODLYtRb0xEGhZQiSNDNQU0INozKcwaAVdy3IQANgA7MzNfFo164z7E7bk0GOS7okA)
```ts
declare const test1: 1
declare const test2: 1 | 2
declare const test3: 2 | 3
declare const test4: number

declare const array: (0 | 1)[]

array.includes(test1) // OK
array.includes(test2) // Argument of type '1 | 2' is not assignable to parameter of type '0 | 1'. Type '2' is not assignable to type '0 | 1'.(2345)
array.includes(test3) // Argument of type '2 | 3' is not assignable to parameter of type '0 | 1'. Type '2' is not assignable to type '0 | 1'.(2345)
array.includes(test4) // Argument of type 'number' is not assignable to parameter of type '0 | 1'.(2345)
```

In this example, the result of `test2` and `test4` is undesired as the test potentially can yield `true` value, much like the following snippet.

```ts
array.findIndex((v) => v === test1) // OK
array.findIndex((v) => v === test2) // OK
array.findIndex((v) => v === test3) // This condition will always return 'false' since the types '0 | 1' and '2 | 3' have no overlap.(2367)
array.findIndex((v) => v === test4) // OK
```

Disputably, some people may even argue that even `test3` should not produce an error. ¯\\_(ツ)_/¯

**Comment by whzx5byb** [comment]
@ackvf See discussion in https://github.com/microsoft/TypeScript/issues/14520.

**Comment by nevnein** [comment]
I hope this is not out of scope, but when one of the mentioned methods is called on a tuple of length `n`, the index parameter could be inferred as `0 | 1 | … | n -1` instead of just `number`. TS already knows the tuple length, and this really could be really useful when accessing other tuples of the same length and `noUncheckedIndexedAccess` is set to `true`.

Compiler settings: default, but with `noUncheckedIndexedAccess` set to `true`
Compiler version: 5.2.2
Example:
```ts
const aTuple = [ "a", "b", "c"] as const
const bTuple = [ "x", "y", "z"] as const

// These are both correctly inferred as exactly 3
const aLength = aTuple.length
const bLength = bTuple.length

type range = 0 | 1 | 2
const possibleIndex = 2 as range // This could be something like getRandomNumber(0,2)
// This is inferred as x | y | z and doesn't include undefined, as expected
const possibleAccess = bTuple[possibleIndex]

// This causes an error, because the return array could contain undefined
const test: string[] = aTuple.map((_, i) => {
  // With noUncheckedIndexedAccess=true access is inferred as possibily undefined, 
  // but TS could/should know that the the index will always be 0 | 1 | 2, as above
  const access = bTuple[i]
  return access
})
```

Here is the [playground](https://www.typescriptlang.org/play?noUncheckedIndexedAccess=true&ssl=19&ssc=3&pln=1&pc=1#code/MYewdgzgLgBAhgFQK4AcA2BTGBeGBtGAIjkIBoiAjMo4QgXXghlEigCgXoYLl0tcChAB7VCAT1EAveo2bhobNgHolMBAAsMELHABOWCiCjq5u-cChoxMAJZgAZhjMYAJrIxC4FqzADMHeVg4ABkMMABzYxx4XkwAOkwI4wDWblCkk1weVHjEyPVFKDEULF04CP4YAAYYAB8YAEY6mAAmFK4UEAgIGwpMAEkwFw9oltkyipgVNXUbJlAkNDcKLAgQAFsMYztwmDQbAGsscK2AJXKXDYA5JHWV3QAKKtIWgEplVQ052yY7R2c3HAmEJmtZ6pJ4EMYJctGAAOSwOzANBIYYwJBDDD2OyuchAmAeEoWVztWCdbq9TAAQWAwC0TCysQweHJPT6GEGwyEdEU0y+8zgSG0THKBLMIF05BWwEF2hgxlKWyQujA8DMcGsCyWcjAUDgdnRmOxYBJnFgUC0UAAXDBoLodngGLhEDkMHF1nAUA8HgB9cg2V44AB8MAA3mwYFNVAB1GxRMAgACqYGAmmARxcnI8rhpdO62CguiQOlp9J+tgcTn0gKYrN6Nh8GOGxtxMAjUe4SFgCAAynJFi4lBB1CABzADgmAO7y9Rwc2aGdYOxcmCThtoeBoScapgrarNJr1Fp4kWGABuGHbZvgpe60WyfDwNh5kf0UGVqq8eYgbAAvq8gA)

**Comment by karlismelderis-mckinsey** [comment]
I would appreciate if Array.includes would behave like this:
```ts
interface Array<T> {
  includes(searchElement: unknown, fromIndex?: number): searchElement is T;
}

interface ReadonlyArray<T> {
  includes(searchElement: unknown, fromIndex?: number): searchElement is T;
}
```

**Comment by icecream17** [comment]
it depends, the most you can guarantee is `searchElement is T ? boolean : false` since even if the type matches it's still possible for includes to return false:

```ts
const abcs: String[] = "abcs".split('')

abcs.includes("d") // > false, even though "d" matches type String
```

**Comment by Retsam** [comment]
@karlismelderis-mckinsey Yeah this has been suggested before e.g. #31018 and there's some issues.  One is, like @icecream17 said it can return false even when `searchElement` actually is `T`- which would require #15048: a false result on the `includes` check can't be used to prove that `searchElement` isn't `T`.

---

But also, I think widening `searchElement` from `T` to `unknown` is problematic for other use cases.  There's two reasons to use `includes`:

1.  You're using the string to check something about the array. e.g. `if(colorArray.includes("red"))`
2. You're using the array to check something about the string. e.g. `if(["red", "blue", "green"].includes(maybeColor))`

The type-guard signature you suggest is useful for #2, but makes #1 worse: if `type Color = "red" | "blue" | "green"` and `colorArray` is `Color[]`, it's not great if `colorArrray.includes("read")` is accepted: that's a typo that's currently caught but wouldn't be if `includes` took `unknown` as its argument.

---

Personally, I suggest making a utility function for #2 and using it in place of `arr.includes(val)`:

```ts
function includes<const S>(haystack: readonly S[], needle: unknown): needle is S {
    const _haystack: readonly unknown[] = haystack;
    return _haystack.includes(needle)
}
```

**Comment by Retsam** [comment]
(For convenience `type Sister = "Louisa" | "Isabella" | "Mirabel"`)

@dsongman Are you saying that the `sisters` array should be typed as `Sister[]` (which is the current type), but somehow remember that it was defined as `satisfies Characters[]` and change the signature of `includes`?  

I don't think there's any plausible mechanism for that - ultimately the methods of arrays are defined by normal type system definitions:

```ts
interface Array<T> {
    includes(searchElement: T, fromIndex?: number): boolean
}
```
In this case `T` is `Sister` and so trying to pass `Characters` as `searchElement` doesn't work, for the reasons specified above, and `satisfies` can't change that.  Either `T` is `Sister` (as in the above code) or you can do `const sisters: Characters[] = /*...*/` to have T` be `Characters`, but I don't think there's a way to have both.

**Comment by dsongman** [comment]
> Are you saying that the sisters array should be typed as Sister[] (which is the current type), but somehow remember that it was defined as satisfies Characters[] and change the signature of includes?

Yeah, that's right. Backing up to make sure I'm stating the problem and not necessarily the solution…

I find that in the application codebases I've worked in, a very frequent issue that comes up is wanting to define a list of values via a union and then iterating over those values in the UI somewhere. If you're okay with duplication, you can make a type that ensures an array contains all values of a union. If you want a single source of truth, though, the general consensus seems to be to define an array of thing things, add `as const` to it, and then you can derive the union via `typeof myThing[number]`.

If, however, you want to define a _subset_ of a union that is iteratable, there isn't a great solution. Using `satisfies SuperType[]` does a lot:
* It ensures that your values conform to `SuperType` as you type them.
* The subtype array or one of its values can be passed to functions that want `SuperType[]` or `SuperType`.  This seems to imply that the `satisfies` predicate holds, since the same is not true for the `as const` version of the same array (see [this playground](https://www.typescriptlang.org/play/?#code/C4TwDgpgBAwgFgQwE4IMbAkgzlAvFAcgEEAjAVwgBsECoAfQgKTMoEsJgb7CiBzMrMAC3AO1oMCAWVYoSVcYQCSWBHMrUFBADIB7MqxUEA3ACgTqHSMFRBSViN4BlAxmxQ8UANon377XoMaABofXwJlVSoNEN9CaVl5EIBdU3NLawsrYGdBTBx8b1j-fUMYsIi1aNC-eMjKAmSoBBxMwVTW4BsEYAMAMzRgHSQQHNd8r2rCXRLgyfCVStmi2rUGkySunqxe9hx4ZAG8zxSzDqgwOxFgfZR0PI8AClREW7GALlgXw+xjgEo8AB8UGeBzu2AAdL0hgBRNBwB5PL5g-64IGtHSUCDgyg6XiI0GuX6-UxnC72a5I1yPEGvTAfG7fLAotHpDFYnF4mnfYlmMlXBlgrAPWz2JwuPI8vkUgl5J7pbLi7CSy7S2nYYXdPoDIYjRVMoxQAD0hs2fV2UAAYgAVADqAEJeSqBa5hcBLmLcj8AAxJX4mKXOzByrKjI4+v0BylBlRbfroHWh72+1LG03bc0AEx0ECwYk6AHdWBmICIoMA4NBkCgQFAALYcOA6DMAfjSWSgBkT1KjSHpPaZby5YI7OAeoEgOl6abjg2GiaZnhEZFrciQG1R0+1c714PsqEoZGLQqHhKMQA)).

But while you can pass the subtype to functions that accept `SuperType`, you can't use `includes` to see if a member of `SuperType` is in the subtype, which seems odd.

**Comment by luukvhoudt** [comment]
`Array.shift()` and `Array.pop()` always return an union of all array item types, even when each array item has its own type. I this case I rather expect to get the type specific to the array item being returned. So for `shift` the first item type, for `pop` the last:

```ts
declare const subject: [string, null, number];

function test_shift_returns_first_type_in_the_array(): void {
    const result = subject.shift() ?? null;
    const expected: string | null = result;
    //    ^ Type 'string | number | null' is not assignable to type 'string | null'.
    //        Type 'number' is not assignable to type 'string'.
    const actual: string | number | null = result;
}

function test_pop_returns_last_type_in_the_array(): void {
    const result = subject.pop() ?? null;
    const expected: number | null = result;
    //    ^ Type 'string | number | null' is not assignable to type 'number | null'.
    //        Type 'string' is not assignable to type 'number'.
    const actual: string | number | null = result;
}
```


[TS Playground](https://www.typescriptlang.org/play/?ts=5.8.3#code/CYUwxgNghgTiAEYD2A7AzgF3mgrgIwCtwMAueAbUxgEsUBzAGnhRwgiZYFs8QYBdANwAoIQDMcKMBmqp4GEJgD6aABbVRGRXAw4Y6RaOowlGAJ4AHEItqKMKq7BhRTACgCUZAG5JqweAG8heGDEVEx4OFwILABebHwiKQA6VXUMd3gAfkzmVghhEND0LBAAD0spEGAyKlo6eAAfXLZ4OMjWDAKQgHpuwoA9eAAVCwQAclr6RtzuXmmWNjH4ajRmJCwoNDRqOhQoPAgEDCQ5UfgJjBoppoWIMaSgnr7CkJHLc64eGCWVtY2tnZ7A5HE5md4XK50e6PYLIYrwKBSHBQCA1S51eY4WYwTEtNoKDrCAC+InEkmksnkSnMSHMWhAOj0aEU0BMo2sKFs9kUjmc7i8Pj8gUKcPC7WirXihGISRp5gy2Wa+RhRXCZQq8mqMy+uIgkvFnRVvQGwzOEIxNyxOsti2WqxQ6wRAN2+0OclBZs+cxtdwehWNL2Cb3Gkyhdr+Tu2LuB7tO4K93z9IVFGyRKLRkMx2N1+oJ0WJQA)

**Comment by icecream17** [comment]
@luukvhoudt

```ts
test_pop_returns_last_type_in_the_array()
test_pop_returns_last_type_in_the_array()
test_pop_returns_last_type_in_the_array() // > results in a string
```
